core/src/main/scala/org/typelevel/paiges/Doc.scala

```diff
@@ -576,19 +576,16 @@ object Doc {
   /**
    * Represents an optimistic rendering (on the left) as well as a
    * fallback rendering (on the right) if the first line of the left
-   * is too long.
+   * is too long.  In `Union(a, b)`, we have the invariants:
    *
-   * There is an additional invariant on Union: `flatten(a) == flatten(b)`.
+   * - `a.flatten == b.flatten`
+   * - `a != b` (otherwise the Union would be redundant)
+   * - `a` is right-associated with respect to `Concat` nodes to maintain efficiency in rendering.
```

Comment:
do we really need this as an invariant? If it is cheap it is nice to have it, but should we really do O(N) work in fill to get it, and I think we are doing that in the current PR.

Maybe we can say that we try, where we can do at most a constant more work, to get things right associated, which allows O(1) work to compute next line.

Maybe you can imagine k-right associated means that you have to recurse at most `k` times on `Concat` nodes to reach a left side that is not a concat node. In which case, `1-right` associated means that if you have a `Concat(a, b)` then `a` is not a Concat. But `2-right` associated means that `Concat(Concat(a, b), c)` is allowed.

In this, maybe we just want `O(1)-right associated`. I think fill does this since it makes Concats out of right associated things. So, maybe 2-right associated is what fill is doing?